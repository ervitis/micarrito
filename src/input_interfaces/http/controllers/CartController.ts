import { createController } from 'awilix-express'
import { BAD_REQUEST, INTERNAL_SERVER_ERROR, OK } from 'http-status-codes'

class CartController {
  private cartService
  private logger

  constructor ({ cartService, logger }) {
    this.cartService = cartService
    this.logger = logger
  }

  async getAll (req, res) {
    try {
      const items = await this.cartService.execute()
      res.status(OK).send({items: items})
    } catch (err) {
      this.logger.error(err)
      res.status(INTERNAL_SERVER_ERROR).send({ error: 'Error retrieving items from carts' })
    }
  }
}

export default createController(CartController)
  .prefix('/cart')
  .get('/items', 'getAll')
