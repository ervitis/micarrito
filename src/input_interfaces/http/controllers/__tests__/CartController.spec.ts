import * as express from 'express'
import { asValue, createContainer } from 'awilix'
import { INTERNAL_SERVER_ERROR } from 'http-status-codes'
import * as bodyParser from 'body-parser'
import { controller, scopePerRequest } from 'awilix-express'
import CartController from '../CartController'
import mockCartService from './__mocks__/mockCartService'
import * as request from 'supertest'

describe('CartController', () => {
  let app, container

  beforeEach(() => {
    app = express()
    container = createContainer()

    app.use((req, res, next) => {
      res.sendInternalServerError = (msg) => () => res.status(INTERNAL_SERVER_ERROR).send(msg)
      res.sendError = (err) => res.status(err.code).send()

      return next()
    })
    app
      .use(bodyParser.json())
      .use(scopePerRequest(container))
      .use(controller(CartController))

    container.register({
      cartService: asValue({})
    })
  })

  afterEach(() => {
    app = null
  })

  test('Get all items', done => {
    const items = [
      { id: 1, productName: 'Knife', quantity: 2, price: 20.20 },
      { id: 2, productName: 'Spoon', quantity: 1, price: 15.0 }
    ]
    const operation = mockCartService('SUCCESS', items)
    container.register({
      cartService: asValue(operation)
    })

    request(app)
      .get('/cart/items')
      .expect(200)
      .end((err, res) => {
        expect(err).toBeFalsy()
        expect(operation.execute).toHaveBeenCalled()
        expect(operation.fns['SUCCESS']).toHaveBeenCalledWith(items)

        done()
      })
  })

  test('Handle error', done => {
    const operation = mockCartService('ERROR', { err: 'Something went wrong' })
    container.register({
      cartService: asValue(operation)
    })

    request(app)
      .get('/cart/items')
      .expect(INTERNAL_SERVER_ERROR)
      .end((err, res) => {
        expect(err).toBeTruthy()
        expect(operation.execute).toHaveBeenCalled()
        expect(operation.fns['ERROR']).toHaveBeenCalled()

        done()
      })
  })
})
