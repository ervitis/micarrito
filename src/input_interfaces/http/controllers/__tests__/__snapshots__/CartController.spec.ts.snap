// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartController Get all items 1`] = `
Array [
  Object {
    "id": 1,
    "price": 20.2,
    "productName": "Knife",
    "quantity": 2,
  },
  Object {
    "id": 2,
    "price": 15,
    "productName": "Spoon",
    "quantity": 1,
  },
]
`;

exports[`CartController Handle error 1`] = `
Object {
  "error": "Error retrieving items from carts",
}
`;
